### PR TITLE
[Draft][WIP]Flex attention

### PR DESCRIPTION
**Pull request by yubofredwang**
## Motivation

Flex Attention allows optimizing speed and memory usage. We should replace the draft model's current naive attention calculation with flex attention.

idea is to instead of using a for loop to calculate attention weights and attention output with previous step's cached hidden states:
```
for i in range(1, lck):
    ki = cache_k[i]
    qi = query_states
    kiq = ki
    attn_weightsi = (qi * kiq).sum(-1) / math.sqrt(self.head_dim)
    attn_weights = torch.cat(
        (attn_weights, attn_weightsi[..., None]), dim=-1
    )
```

Instead we keep a kv cache and keep appending each step's new k and v to it. Then build a custom mask similar to this:

Screenshot 2025-07-31 at 2 30 44 AM

Each step we call flex_attention with the accumulated kv cache.

Initial benchmark result:

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by FlamingoPg** [comment]
Flex Attention look good, it will help save more memory

**Comment by yubofredwang** [comment]
Convergence Test: 3000 x batche size 2.
```
train/acc_0
0.8600095552508316
train/acc_1
0.8408982318008131
train/acc_2
0.8160535113157891
train/acc_3
0.7955088386070193
train/acc_4
0.6196769447581837
train/acc_5
0.5873715116191314
train/acc_6
0.5580029360381749
train/lr
0.00008857775715097407
train/ploss_0
0.3369951844215393
train/ploss_1
0.38721174001693726
train/ploss_2
0.4495728313922882
train/ploss_3
0.4987132251262665
train/ploss_4
0.5543057322502136
train/ploss_5
0.611555814743042
train/ploss_6
0.6813879609107971
```

**Comment by ThomasKluiters** [comment]
When running the Qwen3 MoE example I run into this issue: https://github.com/pytorch/pytorch/pull/160192 using the `aot_eager` backend for compilation I am able to run this on the B200. However, I did not (yet) observe the speed-up or memory benefit.

**Comment by yubofredwang** [comment]
> When running the Qwen3 MoE example I run into this issue: [pytorch/pytorch#160192](https://github.com/pytorch/pytorch/pull/160192) using the `aot_eager` backend for compilation I am able to run this on the B200. However, I did not (yet) observe the speed-up or memory benefit.

I was not able to test on B200. Can you provide a profiling and a benchmark script result?

**Comment by ThomasKluiters** [comment]
> > When running the Qwen3 MoE example I run into this issue: [pytorch/pytorch#160192](https://github.com/pytorch/pytorch/pull/160192) using the `aot_eager` backend for compilation I am able to run this on the B200. However, I did not (yet) observe the speed-up or memory benefit.
> 
> I was not able to test on B200. Can you provide a profiling and a benchmark script result?

Would love to try, however, I cannot seem to increase the context size beyond 1024 before running into OOM when compiling with `aot_eager`! 

I got it working with BS=2 and max_length up to 128k, I do not notice a speed-up but this may just be because of the fact that for B200 no proper optimization exists yet.

**Comment by ThomasKluiters** [comment]
```
PyTorch version: 2.8.0+cu128
CUDA available: True
GPU: NVIDIA B200
GPU memory: 178.3614501953125 GB
Testing sequence lengths: [128, 640, 1152, 1664, 2176, 2688, 3200]

==================================================

=== Benchmarking sdpa ===

Testing sequence length: 128
Warming up...
  Time: 0.306s
  Peak memory: 0.685 GB

Testing sequence length: 640
Warming up...
  Time: 0.356s
  Peak memory: 3.856 GB

Testing sequence length: 1152
Warming up...
  Time: 0.459s
  Peak memory: 10.469 GB

Testing sequence length: 1664
Warming up...
  Time: 0.616s
  Peak memory: 20.334 GB

Testing sequence length: 2176
Warming up...
  Time: 0.627s
  Peak memory: 33.461 GB

Testing sequence length: 2688
Warming up...
  Time: 0.742s
  Peak memory: 49.845 GB

Testing sequence length: 3200
Warming up...
  Time: 0.892s
  Peak memory: 69.490 GB

==================================================

=== Benchmarking flex_attention ===

Testing sequence length: 128
Warming up...
  Time: 0.608s
  Peak memory: 0.837 GB

Testing sequence length: 640
Warming up...
  Time: 0.426s
  Peak memory: 9.810 GB

Testing sequence length: 1152
Warming up...
  Time: 0.636s
  Peak memory: 29.343 GB

Testing sequence length: 1664
Warming up...
  Time: 0.961s
  Peak memory: 59.382 GB

Testing sequence length: 2176
Warming up...
  Time: 1.426s
  Peak memory: 99.930 GB

Testing sequence length: 2688
Warming up...
  Time: 2.038s
  Peak memory: 150.984 GB

Testing sequence length: 3200
```

Odd!

**Comment by yubofredwang** [comment]
> ```
> PyTorch version: 2.8.0+cu128
> CUDA available: True
> GPU: NVIDIA B200
> GPU memory: 178.3614501953125 GB
> Testing sequence lengths: [128, 640, 1152, 1664, 2176, 2688, 3200]
> 
> ==================================================
> 
> === Benchmarking sdpa ===
> 
> Testing sequence length: 128
> Warming up...
>   Time: 0.306s
>   Peak memory: 0.685 GB
> 
> Testing sequence length: 640
> Warming up...
>   Time: 0.356s
>   Peak memory: 3.856 GB
> 
> Testing sequence length: 1152
> Warming up...
>   Time: 0.459s
>   Peak memory: 10.469 GB
> 
> Testing sequence length: 1664
> Warming up...
>   Time: 0.616s
>   Peak memory: 20.334 GB
> 
> Testing sequence length: 2176
> Warming up...
>   Time: 0.627s
>   Peak memory: 33.461 GB
> 
> Testing sequence length: 2688
> Warming up...
>   Time: 0.742s
>   Peak memory: 49.845 GB
> 
> Testing sequence length: 3200
> Warming up...
>   Time: 0.892s
>   Peak memory: 69.490 GB
> 
> ==================================================
> 
> === Benchmarking flex_attention ===
> 
> Testing sequence length: 128
> Warming up...
>   Time: 0.608s
>   Peak memory: 0.837 GB
> 
> Testing sequence length: 640
> Warming up...
>   Time: 0.426s
>   Peak memory: 9.810 GB
> 
> Testing sequence length: 1152
> Warming up...
>   Time: 0.636s
>   Peak memory: 29.343 GB
> 
> Testing sequence length: 1664
> Warming up...
>   Time: 0.961s
>   Peak memory: 59.382 GB
> 
> Testing sequence length: 2176
> Warming up...
>   Time: 1.426s
>   Peak memory: 99.930 GB
> 
> Testing sequence length: 2688
> Warming up...
>   Time: 2.038s
>   Peak memory: 150.984 GB
> 
> Testing sequence length: 3200
> ```
> 
> Odd!

I checked the definition of "aot_eager": uses PyTorch eager mode for AOTAutograd’s extracted forward and backward graphs. This means everything is still eager. The entire flex attention optimization relies on the custom triton kernel being compiled to do sparsed version of flash-attention style attention. Thus the gap is understandable. Need for torch to fully support flex attention with torch.compile on B200.

**Comment by Ximingwang-09** [comment]
Great work! I often encounter OOM issues when training the Qwen3-32B model with 16K context. I'd like to know if this PR is ready - I'll use it for testing.

**Comment by ThomasKluiters** [comment]
> > ```
> > PyTorch version: 2.8.0+cu128
> > CUDA available: True
> > GPU: NVIDIA B200
> > GPU memory: 178.3614501953125 GB
> > Testing sequence lengths: [128, 640, 1152, 1664, 2176, 2688, 3200]
> > 
> > ==================================================
> > 
> > === Benchmarking sdpa ===
> > 
> > Testing sequence length: 128
> > Warming up...
> >   Time: 0.306s
> >   Peak memory: 0.685 GB
> > 
> > Testing sequence length: 640
> > Warming up...
> >   Time: 0.356s
> >   Peak memory: 3.856 GB
> > 
> > Testing sequence length: 1152
> > Warming up...
> >   Time: 0.459s
> >   Peak memory: 10.469 GB
> > 
> > Testing sequence length: 1664
> > Warming up...
> >   Time: 0.616s
> >   Peak memory: 20.334 GB
> > 
> > Testing sequence length: 2176
> > Warming up...
> >   Time: 0.627s
> >   Peak memory: 33.461 GB
> > 
> > Testing sequence length: 2688
> > Warming up...
> >   Time: 0.742s
> >   Peak memory: 49.845 GB
> > 
> > Testing sequence length: 3200
> > Warming up...
> >   Time: 0.892s
> >   Peak memory: 69.490 GB
> > 
> > ==================================================
> > 
> > === Benchmarking flex_attention ===
> > 
> > Testing sequence length: 128
> > Warming up...
> >   Time: 0.608s
> >   Peak memory: 0.837 GB
> > 
> > Testing sequence length: 640
> > Warming up...
> >   Time: 0.426s
> >   Peak memory: 9.810 GB
> > 
> > Testing sequence length: 1152
> > Warming up...
> >   Time: 0.636s
> >   Peak memory: 29.343 GB
> > 
> > Testing sequence length: 1664
> > Warming up...
> >   Time: 0.961s
> >   Peak memory: 59.382 GB
> > 
> > Testing sequence length: 2176
> > Warming up...
> >   Time: 1.426s
> >   Peak memory: 99.930 GB
> > 
> > Testing sequence length: 2688
> > Warming up...
> >   Time: 2.038s
> >   Peak memory: 150.984 GB
> > 
> > Testing sequence length: 3200
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Odd!
> 
> I checked the definition of "aot_eager": uses PyTorch eager mode for AOTAutograd’s extracted forward and backward graphs. This means everything is still eager. The entire flex attention optimization relies on the custom triton kernel being compiled to do sparsed version of flash-attention style attention. Thus the gap is understandable. Need for torch to fully support flex attention with torch.compile on B200.

```
PyTorch version: 2.9.0.dev20250810+cu128
CUDA available: True
GPU: NVIDIA B200
GPU memory: 178.3614501953125 GB
Testing sequence lengths: [128, 640, 1152, 1664, 2176, 2688, 3200]

==================================================

=== Benchmarking sdpa ===

Testing sequence length: 128
Warming up...
  Time: 0.316s
  Peak memory: 0.975 GB

Testing sequence length: 640
Warming up...
  Time: 0.423s
  Peak memory: 7.542 GB

Testing sequence length: 1152
Warming up...
  Time: 0.549s
  Peak memory: 20.759 GB

Testing sequence length: 1664
Warming up...
  Time: 0.742s
  Peak memory: 40.495 GB

Testing sequence length: 2176
Warming up...
  Time: 1.000s
  Peak memory: 66.743 GB

Testing sequence length: 2688
Warming up...
  Time: 1.258s
  Peak memory: 99.513 GB

Testing sequence length: 3200
Warming up...
  Time: 1.615s
  Peak memory: 138.795 GB

==================================================

=== Benchmarking flex_attention ===

Testing sequence length: 128
Warming up...
  Time: 0.326s
  Peak memory: 0.747 GB

Testing sequence length: 640
Warming up...
  Time: 0.422s
  Peak memory: 2.390 GB

Testing sequence length: 1152
Warming up...
  Time: 0.503s
  Peak memory: 4.105 GB

Testing sequence length: 1664
Warming up...
  Time: 0.607s
  Peak memory: 6.246 GB

Testing sequence length: 2176
Warming up...
  Time: 0.729s
  Peak memory: 8.737 GB

Testing sequence length: 2688
Warming up...
  Time: 0.882s
  Peak memory: 11.519 GB

Testing sequence length: 3200
Warming up...
  Time: 1.043s
  Peak memory: 14.588 GB

=== Performance Summary ===
Sequence lengths tested: [128, 640, 1152, 1664, 2176, 2688, 3200]

Speed ratios (Eagle/Flex):
   128: 0.97x
   640: 1.00x
  1152: 1.09x
  1664: 1.22x
  2176: 1.37x
  2688: 1.43x
  3200: 1.55x

Memory ratios (Eagle/Flex):
   128: 1.31x
   640: 3.16x
  1152: 5.06x
  1664: 6.48x
  2176: 7.64x
  2688: 8.64x
  3200: 9.51x
```

Awesome! using (`2.9.0.dev20250810+cu128` and the PR https://github.com/pytorch/pytorch/pull/160192)

**Comment by yubofredwang** [comment]
Training result of 2 epochs of ultrachat + sharegpt @zyksir 
```
eval/epochacc_0
0.747989296913147
eval/epochacc_1
0.7136971354484558
eval/epochacc_2
0.6947001814842224
eval/epochacc_3
0.6795716285705566
eval/epochacc_4
0.6651400327682495
eval/epochacc_5
0.6500007510185242
eval/epochacc_6
0.6327874660491943
eval/epochploss_0
0.664620041847229
eval/epochploss_1
0.7474215030670166
eval/epochploss_2
0.7997110486030579
eval/epochploss_3
0.8433398604393005
eval/epochploss_4
0.8851369023323059
eval/epochploss_5
0.9295921325683594
eval/epochploss_6
0.9807317852973938
train/acc_0
0.7644836262412045
train/acc_1
0.7153652383937465
train/acc_2
0.7178841300782316
train/acc_3
0.7027707799713214
train/acc_4
0.6926952132333813
train/acc_5
0.6788413089687138
train/acc_6
0.6624685130195611
train/epochacc_0
0.747989296913147
train/epochacc_1
0.7136971354484558
train/epochacc_2
0.6947001814842224
train/epochacc_3
0.6795716285705566
train/epochacc_4
0.6651400327682495
train/epochacc_5
0.6500007510185242
train/epochacc_6
0.6327874660491943
train/epochploss_0
0.664620041847229
train/epochploss_1
0.7474215030670166
train/epochploss_2
0.7997110486030579
train/epochploss_3
0.8433398604393005
train/epochploss_4
0.8851369023323059
train/epochploss_5
0.9295921325683594
train/epochploss_6
0.9807317852973938
```

File: specforge/modeling/draft/flex_attention.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    key_cache = torch.randn(
+        B, H, KV_LEN, D, device="cuda", dtype=data_type, requires_grad=True
+    )
+    value_cache = torch.randn(
+        B, H, KV_LEN, D, device="cuda", dtype=data_type, requires_grad=True
+    )
+    seq_lengths = torch.tensor([S], device="cuda", dtype=torch.int32)
+    block_mask = create_block_mask(
+        mask_mod=generate_eagle3_mask(
+            seq_lengths=seq_lengths, Q_LEN=Q_LEN, KV_LEN=KV_LEN, shift_left=128 * 2
```
Comment (FrankLeeeee) [inline specforge/modeling/draft/flex_attention.py]: This test should be better moved to `tests/test_flex_attention.py`

File: scripts/train_eagle3_online.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ def parse_args():
         default=20,
         help="Timeout for collective communication in minutes",
     )
+    parser.add_argument("--attention-backend", type=str, default="sdpa")
```
Comment (FrankLeeeee) [inline scripts/train_eagle3_online.py]: if flex is better, should we just make it default?
